File: leetcode/C++/unique-paths-ii.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obs) {
        int h=obs.size(), w=obs[0].size();
        vector<int> prev(w,0),cur(w,0);
        if(obs[0][0]==0) cur[0]=1;
        for(int i=0; i<h; ++i){
            for(int j=0; j<w; ++j){
                if(i+j==0 || obs[i][j]) continue;
                if(i) cur[j]+=prev[j];
                if(j) cur[j]+=cur[j-1];
            }
            for(int j=0; j<w; ++j) {prev[j]=cur[j]; cur[j]=0;}
        }
        return prev[w-1];
    }
};

File: leetcode/Python/unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        h, w = len(obstacleGrid), len(obstacleGrid[0])
        cur = prev = [0 for i in range(w)]
        if obstacleGrid[0][0] == 0:
            cur[0] = 1
        for i in range(h):
            for j in range(w):
                if obstacleGrid[i][j]:
                    continue
                if i:
                    cur[j]+=prev[j]
                if j:
                    cur[j]+=cur[j-1]
            prev = cur
            cur = [0 for j in range(w)]
        return prev[w-1]
